### PR TITLE
Add cat021

### PR DESCRIPTION
**Pull request by DabeerM**

### CODE REVIEW

**Comment by david-stubbersfield** [comment]
Fix case TableLW filename.

File: transform/cat021_v10.go [hunk 1]
```diff
@@ -0,0 +1,1162 @@
...
+	"strings"
+
+	"github.com/mokhtarimokhtar/goasterix"
+)
+
+const (
+	BYTESIZE = 8
+)
+
+var ErrTypeUnknown021 = errors.New("[ASTERIX Error CAT021] Message TYPE Unknown")
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go]: not used anywhere. Remove.

File: transform/cat021_v10.go [hunk 2]
```diff
@@ -0,0 +1,1162 @@
...
+	Latitude  float32 `json:"latitude,omitempty"`
+	Longitude float32 `json:"longitude,omitempty"`
+}
+
+type GeometricHeight struct {
+	Height      float64 `json:"height,omitempty"`
+	GreaterThan bool    `json:"greaterthan,omitempty"`
+}
+
+type AirSpeed struct {
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go:25]: Only TrueAirspeed in the spec. Why Airspeed?

File: transform/cat021_v10_test.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+import (
+	"math"
+	"testing"
+)
+
+func equalWithinErrorBounds(actualValue float64, targetValue float64, epsilon float64) bool {
+	return math.Abs(targetValue-actualValue) < epsilon
+}
+
+func checkEqualLatLong(resultCoordinates WGS84Coordinates, actualCoordinates WGS84Coordinates, epsilon float64) bool {
```
Comment (laurence-aung) [inline transform/cat021_v10_test.go]: did you see those functions somewhere else in the repo or are these your own? some engineers prefer unit tests to be completely isolated from each other. they don't prefer share functions like this in unit tests. from what I look through the code base, I can't find any private shared functions too. (personally I don't mind) but when contributing to a public repo, probably not a good idea.

**Comment by laurence-aung** [comment]
every file should end with new line character. applies to every file that missing line ending.

File: transform/cat021_v10.go [hunk 4]
```diff
@@ -0,0 +1,1162 @@
...
+}
+type TargetReportDescriptor struct {
+	ATP string             `json:"atp,omitempty"`
+	ARC string             `json:"arc,omitempty"`
+	RC  string             `json:"rc,omitempty"`
+	RAB string             `json:"rab,omitempty"`
+	FX  *FirstExtensionTRD `json:"fx,omitempty"`
+}
+
+type NICVersion2OrHigher struct {
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go]: underscore between NIC and Version?

File: transform/cat021_v10.go [hunk 5]
```diff
@@ -0,0 +1,1214 @@
...
+
+type TargetStatus struct {
+	ICF  string `json:"icf,omitempty"`
+	LNAV string `json:"lnav,omitempty"`
+	ME   bool   `json:"me,omitempty"`
+	PS   string `json:"ps,omitempty"`
+	SS   string `json:"ss,omitempty"`
+}
+
+type VerticalRate struct {
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go:125]: Geo or Baro? Should be both right?

Comment (david-stubbersfield) [reply transform/cat021_v10.go:125]: I think you're using a generic type for both right? Smart :)

File: transform/cat021_v10.go [hunk 6]
```diff
@@ -0,0 +1,1214 @@
...
+func (data *Cat021Model) write(rec goasterix.Record) {
+	for _, item := range rec.Items {
+		switch item.Meta.FRN {
+		case 1:
+			var payload [2]byte
+			copy(payload[:], item.Fixed.Data[:])
+			tmp, _ := sacSic(payload)
+			data.DataSourceIdentification = &tmp
+		case 2:
+			// TODO: Write unit tests
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go]: confirm TODO still relevant.

File: transform/cat021_v10.go [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+			// Do stuff
+		case 49:
+			// Do stuff
+		}
+	}
+}
+
+// TODO: Refactor to cover for arbitrary number of extensions (currently only covers
+//       two as that's explicitly in the spec)
+func getTargetReportDescriptor(cp goasterix.Compound) TargetReportDescriptor {
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go]: notice functions in other file don't use 'get'. Probs remove.

File: transform/cat021_v10.go [hunk 8]
```diff
@@ -0,0 +1,1214 @@
...
+			fx1.FX = fx2
+		}
+
+		qi.FX = fx1
+	}
+
+	return *qi
+}
+
+func wgs84Coordinates(data []byte) WGS84Coordinates {
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go:627]: check equivalent implementation in cat062 as to whether more suitable.

File: transform/cat021_v10.go [hunk 9]
```diff
@@ -0,0 +1,1214 @@
...
+}
+
+func getTrueAirSpeed(data [2]byte) TrueAirSpeed {
+	return TrueAirSpeed{
+		RE:    int(data[0] & 0x80),
+		Speed: int16(uint32(data[0]&0x7F)<<BYTESIZE + uint32(data[1]&0xFF)),
+	}
+}
+
+// TODO: Double-check this...
```
Comment (david-stubbersfield) [inline transform/cat021_v10.go]: have you done this? Reminder.

File: transform/common.go [hunk 10]
```diff
@@ -37,6 +46,31 @@ func timeOfDay(data [3]byte) (tod float64, err error) {
 	return tod, nil
 }
 
+// timeOfDay returns a float64 in second (1 bit = 1/2^30 s)
+// Absolute time stamping expressed as Co-ordinated Universal Time (UTC).
```
Comment (david-stubbersfield) [inline transform/common.go:50]: no '-'

File: transform/cat021_v10.go (hunk 2 above)
Comment (DabeerM) [reply transform/cat021_v10.go:25]: On page 83 of the spec (https://www.eurocontrol.int/sites/default/files/2021-02/asterix-adsbtr-cat021-part12-v2-5.pdf), the FRNs for both are listed

File: transform/cat021_v10.go (hunk 4 above)
Comment (DabeerM) [reply transform/cat021_v10.go]: Agreed and added

File: transform/cat021_v10.go (hunk 5 above)
Comment (DabeerM) [reply transform/cat021_v10.go:125]: Yep, when looking at the spec, they're both the same effectively :p

File: transform/cat021_v10.go (hunk 8 above)
Comment (DabeerM) [reply transform/cat021_v10.go:627]: no can do - different precision (and size)

File: transform/common.go (hunk 10 above)
Comment (DabeerM) [reply transform/common.go:50]: done

File: transform/cat021_v10.go (hunk 1 above)
Comment (DabeerM) [reply transform/cat021_v10.go]: done

File: transform/cat021_v10_test.go (hunk 3 above)
Comment (DabeerM) [reply transform/cat021_v10_test.go]: moved it to common.go

File: transform/cat021_v10.go (hunk 7 above)
Comment (DabeerM) [reply transform/cat021_v10.go]: done

File: transform/cat021_v10.go (hunk 9 above)
Comment (DabeerM) [reply transform/cat021_v10.go]: ye

File: transform/cat021_v10.go (hunk 6 above)
Comment (DabeerM) [reply transform/cat021_v10.go]: done

**Pull request merged**